pytorch_lightning/plugins/training_type/fully_sharded.py

```diff
@@ -0,0 +1,213 @@
+                Using ``min_num_params`` to determine the amount of parameters to wrap at a time.
+                (default: False)
+
+            min_num_params: Number of parameters to wrap when using FairScale ``auto_wrap``.
+                (default: 1e8)
+
+        """
+        if not _FAIRSCALE_FULLY_SHARDED_AVAILABLE:
+            raise MisconfigurationException(
+                "Full Sharded Training is not available. Install the latest FairScale via `pip install fairscale -U`"
+            )
+
+        if sync_batchnorm:
+            raise MisconfigurationException("Currently sync batch norm is not supported by Full Sharded Training.")
+        super().__init__(parallel_devices, num_nodes, cluster_environment, sync_batchnorm)
+        self.cpu_offload = cpu_offload
+        self.move_grads_to_cpu = move_grads_to_cpu
+        self.flatten_parameters = flatten_parameters
+        self.reshard_after_forward = reshard_after_forward
+        self.fp32_reduce_scatter = fp32_reduce_scatter
+        self.compute_dtype = compute_dtype
+        self.bucket_cap_mb = bucket_cap_mb
+        self.automatic_module_wrap = automatic_module_wrap
+        self.min_num_params = min_num_params
+        self._process_group = None
+
+    @property
+    def process_group(self):
+        if self._process_group is None:
+            self._process_group = torch.distributed.new_group()
+        return self._process_group
+
+    def setup_distributed(self):
+        super().setup_distributed()
+        if self.root_device.type == "cuda":
+            torch.cuda.set_device(self.root_device)
+
+    @contextlib.contextmanager
+    def model_sharded_context(self) -> Generator:
+        precision = self.lightning_module.trainer.precision
+
+        def wrap_policy(*args, **kwargs):
+            return default_auto_wrap_policy(*args, **kwargs, min_num_params=self.min_num_params)
+
+        with enable_wrap(
+            wrapper_cls=FullyShardedDataParallel,
+            auto_wrap_policy=wrap_policy,
+            process_group=self.process_group,
+            cpu_offload=self.cpu_offload,
+            move_grads_to_cpu=self.move_grads_to_cpu,
+            flatten_parameters=self.flatten_parameters,
+            mixed_precision=precision == "mixed",
+            reshard_after_forward=self.reshard_after_forward,
+            fp32_reduce_scatter=self.fp32_reduce_scatter,
+            compute_dtype=self.compute_dtype,
+            bucket_cap_mb=self.bucket_cap_mb,
+        ):
+            yield
+
+    def configure_ddp(self):
+        with self.model_sharded_context():
+            if self.automatic_module_wrap and not self._model_has_nested_fsdp():
+                self.model = auto_wrap(LightningFullyShardedModule(self.model))
+                if not isinstance(self.model, FullyShardedDataParallel):
+                    self.model = wrap(self.model)
+            else:
+                self.model = wrap(LightningFullyShardedModule(self.model))
```

Comment:
if manually wrapping contents inside the lightning module, is this final outer layer wrap needed? or could we defer this to the user in the lightning module too?

then we could not wrap model in the dummy `LightningFullyShardedModule` to map forward to one of the step functions. would it also mean users don't have to refer to self.trainer.model inside of the lightning module? 

would this avoid the parameter flattening issue across stages?

Reply:
Ah I understand! Theoretically I think so, since then we're just using the LM as a wrapper. So the cases I see:

1. User wraps nothing, expects module to be wrapped by Lightning, and potentially `auto_wrap` to handle recursive wrapping
2. User wraps **some** of the layers in `configure_sharded_model` but then expects all other layers to be included in a higher wrapper class (wrap the entire LM)
3. User wraps **all** of the layers in `configure_sharded_model`, doesn't require any high level wrapping


Solutions
1. This should be default behaviour, i.e `plugins=fsdp` or `plugins=fsdp_auto_wrap`
2. This should be the same as 1., i.e `plugins=fsdp` or `plugins=fsdp_auto_wrap`
3. This could be `plugins=fsdp_manual` where we do not wrap the highest level module, allowing the user to do whatever they'd like in `configure_optimizers`.

In either case, it's important to fix the flattening issue for 1. and 2. which for most users trying out will be the first step I think. Thoughts @ananthsub?

Reply:
Yes exactly those 3 cases! 

To unblock the initial integration, I was wondering if we should start with option #3 to unblock power users in release candidates with the caveat that they are responsible for the full wrapping. Maybe this could can be option on the plugin as to whether the outer wrap on lightning module needs to be applied in order to distinguish between cases 2 and 3. 

Completely agreed with you that most users will opt for cases 1 and 2, so we'll need to figure out the parameter flattening, whether in lightning or fairscale, but wanted to offer this as one way we could sequence these cases

Reply:
I started making changes to see if any issues arise with case 3 and a few observation:

The user still has to define a single model, may it be a Module containing modules in a sequential wrapper, or just defining their own model structure defining a forward function. This means `self.model` will still probably be required in every case for FSDP to work in `configure_optimizers`.

I also ran into an issue where clipping grad norms which in manual mode cannot be handled automatically, as we do not wrap the model:
```python
class FullyShardedNativeMixedPrecisionPlugin(ShardedNativeMixedPrecisionPlugin):
    """Mixed Precision for Full Sharded Training"""

    def clip_gradients(`
        self,
        optimizer: 'Optimizer',
        clip_val: Union[int, float],
        gradient_clip_algorithm: GradClipAlgorithmType = GradClipAlgorithmType.NORM,
        model: Optional[Module] = None
    ) -> None:
        # Model manages clipping of gradients
        model = cast(FullyShardedDataParallel, model)
        # todo: expose norm type once precision plugin supports this.
        model.clip_grad_norm_(clip_val, norm_type=2.0) # This breaks
```

A potential solution albeit not as elegant as I'd like, would be to go through the immediate children of the LightningModule, find the root FSDP module and call `clip_grad_norm_` on it. I assume this will be a negligible cost added on top of the training loop but what are your thoughts @ananthsub?